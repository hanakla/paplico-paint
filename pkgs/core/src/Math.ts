import { type mat4 } from 'gl-matrix'

type Interpolator = (a: number, b: number, ratio: number) => number

export const lerp = (a: number, b: number, t: number) => {
  return (1 - t) * a + t * b
}

const linearInterpolation = (a: number, b: number, ratio: number) => {
  return a + (b - a) * ratio
}

const getArrayValueClamped = (array: number[], index: number) => {
  // prettier-ignore
  return index < 0 ? array[0] :
    index >= array.length ? array[array.length - 1] :
    array[index]
}

const clampIndex = (array: any[], index: number) => {
  // prettier-ignore
  return index < 0 ? 0 :
    index >= array.length ? array.length - 1 :
    index
}

// from: https://github.com/toji/gl-matrix/blob/master/src/mat4.js
export class Matrix4 {
  // prettier-ignore
  protected matrix: mat4 = [
    1, 0, 0, 0,
    0, 1, 0, 0,
    0, 0, 1, 0,
    0, 0, 0, 1
  ]

  public translate(...vector: [x: number, y: number, z: number]): this {
    const out = this.matrix,
      a = this.matrix.slice(),
      v = vector

    let x = v[0],
      y = v[1],
      z = v[2]
    let a00, a01, a02, a03
    let a10, a11, a12, a13
    let a20, a21, a22, a23

    if (a === out) {
      out[12] = a[0] * x + a[4] * y + a[8] * z + a[12]
      out[13] = a[1] * x + a[5] * y + a[9] * z + a[13]
      out[14] = a[2] * x + a[6] * y + a[10] * z + a[14]
      out[15] = a[3] * x + a[7] * y + a[11] * z + a[15]
    } else {
      a00 = a[0]
      a01 = a[1]
      a02 = a[2]
      a03 = a[3]
      a10 = a[4]
      a11 = a[5]
      a12 = a[6]
      a13 = a[7]
      a20 = a[8]
      a21 = a[9]
      a22 = a[10]
      a23 = a[11]

      out[0] = a00
      out[1] = a01
      out[2] = a02
      out[3] = a03
      out[4] = a10
      out[5] = a11
      out[6] = a12
      out[7] = a13
      out[8] = a20
      out[9] = a21
      out[10] = a22
      out[11] = a23

      out[12] = a00 * x + a10 * y + a20 * z + a[12]
      out[13] = a01 * x + a11 * y + a21 * z + a[13]
      out[14] = a02 * x + a12 * y + a22 * z + a[14]
      out[15] = a03 * x + a13 * y + a23 * z + a[15]
    }

    return this
  }

  public rotateZ(rad: number): this {
    const out = this.matrix,
      a = this.matrix.slice()

    let s = Math.sin(rad)
    let c = Math.cos(rad)
    let a00 = a[0]
    let a01 = a[1]
    let a02 = a[2]
    let a03 = a[3]
    let a10 = a[4]
    let a11 = a[5]
    let a12 = a[6]
    let a13 = a[7]

    if (a !== out) {
      // If the source and destination differ, copy the unchanged last row
      out[8] = a[8]
      out[9] = a[9]
      out[10] = a[10]
      out[11] = a[11]
      out[12] = a[12]
      out[13] = a[13]
      out[14] = a[14]
      out[15] = a[15]
    }

    // Perform axis-specific matrix multiplication
    out[0] = a00 * c + a10 * s
    out[1] = a01 * c + a11 * s
    out[2] = a02 * c + a12 * s
    out[3] = a03 * c + a13 * s
    out[4] = a10 * c - a00 * s
    out[5] = a11 * c - a01 * s
    out[6] = a12 * c - a02 * s
    out[7] = a13 * c - a03 * s

    return this
  }

  public toArray() {
    return this.matrix.slice(0)
  }
}

export const interpolateMap = (
  numbers: number[],
  interpolate: Interpolator = linearInterpolation
) => {
  const fn = (t: number) => {
    const pos = t * (numbers.length - 1)
    return fn.byIndex(pos)
  }

  fn.byIndex = (idx: number) => {
    const rate = idx - Math.trunc(idx)
    const index = Math.trunc(idx)

    // prettier-ignore
    const a = numbers[clampIndex(numbers, index)]
    const b = numbers[clampIndex(numbers, index + 1)]

    return interpolate(a, b, rate)
  }

  return fn
}

export const interpolateMapObject = <T>(
  array: T[],
  picker: (index: number, array: T[]) => number,
  interpolate: Interpolator = linearInterpolation
) => {
  return (t: number) => {
    const pos = t * (array.length - 1)
    const rate = pos - Math.trunc(pos)
    const index = Math.trunc(pos)

    const a = picker(clampIndex(array, index), array)
    const b = picker(clampIndex(array, index + 1), array)

    return interpolate(a, b, rate)
  }
}

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest!

  describe('interpolateMap', () => {
    it('with t', () => {
      const map = interpolateMap([0, 1, 2])

      expect(map(0)).toBe(0)
      expect(map(0.25)).toBe(0.5)
      expect(map(0.5)).toBe(1)
      expect(map(1)).toBe(2)
    })

    it('with idx', () => {
      const map = interpolateMap([0, 1, 2])

      expect(map.byIndex(0)).toBe(0)
      expect(map.byIndex(0.5)).toBe(0.5)
      expect(map.byIndex(1)).toBe(1)
      expect(map.byIndex(1.5)).toBe(1.5)
      expect(map.byIndex(2)).toBe(2)
    })
  })

  describe('Math', () => {
    describe('clampIndex', () => {
      it('should clamp index', () => {
        const array = [0, 1]

        expect(clampIndex(array, -1)).toBe(0)
        expect(clampIndex(array, 0)).toBe(0)
        expect(clampIndex(array, 1)).toBe(1)
        expect(clampIndex(array, 2)).toBe(1)
      })
    })

    describe('interpolateMapObject', () => {
      it('should interpolate', () => {
        const array = [0, 1, 2]
        const intelate = interpolateMapObject(array, (i) => i)

        expect(intelate(0)).toBe(0)
        expect(intelate(0.5)).toBe(1)
        expect(intelate(1)).toBe(2)
      })
    })
  })
}
