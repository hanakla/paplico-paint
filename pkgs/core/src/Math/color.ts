// @author GPT-4
export function hsvToRgb(
  h: number,
  s: number,
  v: number
): [number, number, number] {
  let r: number, g: number, b: number

  const i: number = Math.floor(h * 6)
  const f: number = h * 6 - i
  const p: number = v * (1 - s)
  const q: number = v * (1 - f * s)
  const t: number = v * (1 - (1 - f) * s)

  switch (i % 6) {
    case 0:
      r = v
      g = t
      b = p
      break
    case 1:
      r = q
      g = v
      b = p
      break
    case 2:
      r = p
      g = v
      b = t
      break
    case 3:
      r = p
      g = q
      b = v
      break
    case 4:
      r = t
      g = p
      b = v
      break
    case 5:
      r = v
      g = p
      b = q
      break
    default:
      throw new Error('Invalid HSV color')
  }

  return [r, g, b]
}

// @author GPT-4
export function rgbToHsv(
  r: number,
  g: number,
  b: number
): [number, number, number] {
  // find the maximum and minimum values of the RGB components
  const max = Math.max(r, g, b)
  const min = Math.min(r, g, b)

  let h = 0
  let s = 0
  let v = max

  const delta = max - min

  // calculate the saturation of the color
  if (max !== 0) {
    s = delta / max
  }

  // calculate the hue of the color
  if (delta !== 0) {
    if (max === r) {
      h = (g - b) / delta + (g < b ? 6 : 0)
    } else if (max === g) {
      h = (b - r) / delta + 2
    } else {
      h = (r - g) / delta + 4
    }
    h /= 6
  }

  // return the hue, saturation, and value of the color as an array
  // hue is in the range [0, 1]
  // saturation is in the range [0, 1]
  // value is in the range [0, 1]
  return [h, s, v]
}
