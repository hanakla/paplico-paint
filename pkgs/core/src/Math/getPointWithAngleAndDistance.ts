export function getPointWithAngleAndDistance({
  angle,
  distance,
  base = { x: 0, y: 0 },
}: {
  /** radian number */
  angle: number
  distance: number
  base?: { x: number; y: number }
}) {
  return {
    x: base.x + Math.cos(angle) * distance,
    y: base.y + Math.sin(angle) * distance,
  }
}
