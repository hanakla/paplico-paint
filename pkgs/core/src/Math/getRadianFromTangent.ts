import { getTangent } from './getTangent'

export function getRadianFromTangent(
  x1: number,
  y1: number,
  x2: number,
  y2: number,
) {
  const vector = getTangent(x1, y1, x2, y2)
  const result = Math.atan2(vector.x, vector.y)
  return Number.isNaN(result) ? 0 : result
}
