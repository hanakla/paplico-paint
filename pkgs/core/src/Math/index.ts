export { hsvToRgb, rgbToHsv } from './color'
export { distance2D } from './space'
export { Matrix4, type Vector3Tuple } from './matrix4'
export {
  createNumSequenceMap,
  createObjectSequenceMap,
  mapLinear,
  lerp,
  clampNumInLength,
} from './interpolation'
export { createSeededRandom } from './random'
export { clamp, radToDeg, degToRad } from '@/utils/math'
export { getRadianFromTangent } from './getRadianFromTangent'
export { getTangent } from './getTangent'
export { getPointWithAngleAndDistance } from './getPointWithAngleAndDistance'
