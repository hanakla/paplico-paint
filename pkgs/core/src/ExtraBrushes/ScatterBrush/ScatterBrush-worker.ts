import { pointsToSVGCommandArray } from '@/Engine/VectorUtils'
import { type BrushLayoutData } from '@/index'
import { getRadianFromTangent } from '@/StrokingHelper'
import { Matrix4, mapLinear } from '@/Math'
import { indexedPointAtLength } from '@/fastsvg/IndexedPointAtLength'
import { type VectorPath } from '@/Document/LayerEntity/VectorPath'

export type Payload =
  | { type: 'warming' }
  | { type: 'aborted'; id: string }
  | {
      id: string
      type: 'getPoints'
      path: VectorPath
      destSize: { width: number; height: number }
      brushSize: number
      scatterRange: number
      scatterScale: number
      inOutInfluence: number
      inOutLength: number
    }

export type GetPointWorkerResponse = {
  id: string
  type: 'getPoints'
  matrices: Array<number[] | Float32Array>
  lengths: number[]
  totalLength: number
  bbox: { left: number; top: number; right: number; bottom: number } | null
  _internals: any
}

export type WorkerResponse =
  | { type: 'warming' }
  | { type: 'aborted'; id: string }
  | GetPointWorkerResponse

const abortedTasks = new Set<string>()
const queue: Payload[] = []
const queueResult = new Map<string, WorkerResponse>()

const handleMessage = (
  handler: (event: MessageEvent<Payload>) => Promise<WorkerResponse | undefined>
) => {
  return async (event: MessageEvent<Payload>) => {
    const result = await handler(event)
    if (result != null) {
      self.postMessage(result)
    }
  }
}

if (typeof process === 'undefined' || process.env.NODE_ENV !== 'test') {
  self.onmessage = handleMessage(async ({ data }) => {
    switch (data.type) {
      case 'warming': {
        return { type: 'warming' }
      }

      case 'aborted': {
        // console.log('receive aboterd on worker')
        abortedTasks.add(data.id)
        break
      }

      case 'getPoints': {
        queue.push(data)

        return await new Promise((r) => {
          setTimeout(function waitResult() {
            const result = queueResult.get(data.id)

            if (result != null) {
              queueResult.delete(data.id)
              return r(result)
            }

            setTimeout(waitResult)
          })
        })
      }

      default:
        break
    }
  })

  setTimeout(async function processQueue() {
    const task = queue.shift()

    if (task == null || task.type !== 'getPoints') {
      setTimeout(processQueue)
      return
    }

    try {
      const res = await processInput(task)
      queueResult.set(task.id, res)
    } finally {
      setTimeout(processQueue)
    }
  })
}

export async function processInput(data: Payload): Promise<WorkerResponse> {
  if (data.type !== 'getPoints') {
    throw new Error(`Invalid payload type: ${data.type}}`)
  }

  if (abortedTasks.has(data.id)) return { type: 'aborted', id: data.id }

  const {
    id,
    path,
    destSize,
    brushSize,
    scatterRange,
    scatterScale,
    inOutInfluence,
    inOutLength
  } = data

  const positions: [number, number][] = []
  const lengths: number[] = []
  const matrices: (number[] | Float32Array)[] = []
  const bbox: BrushLayoutData['bbox'] = {
    left: 0,
    top: 0,
    right: 0,
    bottom: 0
  }

  // const scattered = scatterPlot(path, {
  //   counts: path.points.length * 300,
  //   scatterRange: 2,
  //   scatterScale: 1,
  // })

  // console.time('build pal')

  const pal = indexedPointAtLength(
    pointsToSVGCommandArray(path.points, path.closed)
  )
  // console.timeEnd('build pal')
  const totalLen = pal.totalLength

  if (totalLen === 0) {
    return {
      id,
      type: 'getPoints',
      matrices: null,
      lengths: [],
      bbox: null,
      totalLength: 0
    }
  }

  const step = 1

  let count = 0
  let times = 0
  let requestAts: number[] = []

  for (let len = 0; len <= totalLen; len += step) {
    requestAts.push(len)
    requestAts.push(len + 0.01)
  }

  requestAts.push(totalLen)

  const points = pal.atBatch(requestAts)

  for (let idx = 0, l = points.length; idx < l; idx += 1) {
    // wait for tick (for receiving abort message from main thread)
    await Promise.resolve()

    if (abortedTasks.has(id)) {
      // console.info('aboterd on worker')
      abortedTasks.delete(id)
      return { id, type: 'aborted' }
    }

    const len = points[idx].length
    const [x, y] = points[idx].pos //pal.at(len)
    const next = points[idx + 1]?.pos ?? points[idx]?.pos //pal.at(len + 0.01, { seek: false })

    const rad = getRadianFromTangent({ x, y }, { x: next[0], y: next[1] })

    // const ypos =

    // if len in inOutLength from start or end, scale by inOutInfluence
    // prettier-ignore
    // const inoutScale = inOutLength === 0 ? 1
    //   : len <= inOutLength ? inOutInfluence * len / inOutLength
    //   : len >= totalLen - inOutLength ? inOutInfluence * (totalLen - len) / inOutLength
    //   : 1;
    const inoutScale = inOutLength === 0 ? 1
      : len <= inOutLength ? inOutInfluence * len / inOutLength
      : len >= totalLen - inOutLength ? inOutInfluence * (totalLen - len) / inOutLength
      : 1;

    const matt4 = new Matrix4().translate(
      // x,
      // y,
      mapLinear(
        x,
        [0, destSize.width],
        [-destSize.width / 2, destSize.width / 2]
      ),
      mapLinear(
        y,
        [0, destSize.height],
        [destSize.height / 2, -destSize.height / 2]
      ),
      0
    )
    // .scale([brushSize * inoutScale, brushSize * inoutScale, 1])
    // .rotateZ(rad)

    matrices.push(matt4.toArray())
    lengths.push(len)
    positions.push([x, y])

    bbox.left = Math.min(bbox.left, x)
    bbox.top = Math.min(bbox.top, y)
    bbox.right = Math.max(bbox.right, x)
    bbox.bottom = Math.max(bbox.bottom, y)

    count++
  }

  if (abortedTasks.has(id)) {
    abortedTasks.delete(id)
    return { id, type: 'aborted' }
  }

  return {
    type: 'getPoints',
    id,
    bbox,
    lengths,
    totalLength: totalLen,
    matrices,
    _internals: {
      requestAts,
      positions
    }
  }
}
