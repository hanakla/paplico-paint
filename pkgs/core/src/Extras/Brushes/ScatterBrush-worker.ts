import { pointsToSVGCommandArray } from '@/Engine/VectorUtils'
import { type BrushLayoutData } from '@/index'
import { getRadianFromTangent } from '@/StrokeHelper'
import { Matrix4 } from '@/Math'
import { indexedPointAtLength } from '@/fastsvg/IndexedPointAtLength'
import { type VectorPath } from '@/Document/LayerEntity/VectorPath'

export type Payload =
  | { type: 'warming' }
  | { type: 'aborted'; id: string }
  | {
      id: string
      type: 'getPoints'
      path: VectorPath
      destSize: { width: number; height: number }
      brushSize: number
      scatterRange: number
      scatterScale: number
      inOutInfluence: number
      inOutLength: number
    }

export type GetPointWorkerResponse = {
  id: string
  type: 'getPoints'
  matrices: Array<number[] | Float32Array>
  lengths: number[]
  totalLength: number
  bbox: { left: number; top: number; right: number; bottom: number } | null
}

export type WorkerResponse =
  | { type: 'warming' }
  | { type: 'aborted'; id: string }
  | GetPointWorkerResponse

const abortedTasks = new Set<string>()

const handleMessage = (
  handler: (event: MessageEvent<Payload>) => Promise<WorkerResponse | undefined>
) => {
  return async (event: MessageEvent<Payload>) => {
    const result = await handler(event)
    if (result != null) {
      self.postMessage(result)
    }
  }
}

self.onmessage = handleMessage(async ({ data }) => {
  switch (data.type) {
    case 'warming': {
      return { type: 'warming' }
    }

    case 'aborted': {
      abortedTasks.add(data.id)
      break
    }

    case 'getPoints': {
      const {
        id,
        path,
        destSize,
        brushSize,
        scatterRange,
        scatterScale,
        inOutInfluence,
        inOutLength,
      } = data

      const positions: number[] = []
      const lengths: number[] = []
      const matrices: (number[] | Float32Array)[] = []
      const bbox: BrushLayoutData['bbox'] = {
        left: 0,
        top: 0,
        right: 0,
        bottom: 0,
      }

      // const scattered = scatterPlot(path, {
      //   counts: path.points.length * 300,
      //   scatterRange: 2,
      //   scatterScale: 1,
      // })

      const pal = indexedPointAtLength(
        pointsToSVGCommandArray(path.points, path.closed)
      ).getSequencialReader()

      // const seqPal = pal.getSequencialReader()
      const totalLen = pal.totalLength

      if (totalLen === 0) {
        return { id, type: 'getPoints', points: null, bbox: null }
      }

      const step = 500 / totalLen

      for (let len = 0; len <= totalLen; len += step) {
        // wait for tick (for receiving abort message from main thread)
        await Promise.resolve()

        if (abortedTasks.has(id)) {
          abortedTasks.delete(id)
          return { id, type: 'aborted' }
        }

        const [x, y] = pal.at(len)
        const next = pal.at(len + 0.01, { seek: false })

        const rad = getRadianFromTangent({ x, y }, { x: next[0], y: next[1] })

        const ypos = y / destSize.height

        // if len in inOutLength from start or end, scale by inOutInfluence
        // prettier-ignore
        const inoutScale =
          len <= inOutLength ? inOutInfluence * len / inOutLength
          : len >= totalLen - inOutLength ? inOutInfluence * (totalLen - len) / inOutLength
          : 1

        const matt4 = new Matrix4()
          .translate(
            x - destSize.width / 2,
            (1 - ypos) * destSize.height - destSize.height / 2,
            0
          )
          .scale([brushSize * inoutScale, brushSize * inoutScale, 1])
          .rotateZ(rad)

        // _mat4.fromArray(matt4.toArray())
        matrices.push(matt4.toArray())
        lengths.push(len)
        positions.push([x, y])

        bbox.left = Math.min(bbox.left, x)
        bbox.top = Math.min(bbox.top, y)
        bbox.right = Math.max(bbox.right, x)
        bbox.bottom = Math.max(bbox.bottom, y)
      }

      if (abortedTasks.has(id)) {
        abortedTasks.delete(id)
        return { id, type: 'aborted' }
      }

      return {
        type: 'getPoints',
        id,
        bbox,
        lengths,
        totalLength: totalLen,
        matrices,
      }
    }

    default:
      break
  }
})
