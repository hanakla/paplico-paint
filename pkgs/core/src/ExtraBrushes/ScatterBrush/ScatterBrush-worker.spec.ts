import { Matrix4 } from '@/Math'
import { processInput } from './ScatterBrush-worker'
import { indexedPointAtLength } from '@/StrokingHelper'

describe('ScatterBrush-worker', () => {
  it('test', () => {
    const indexed = indexedPointAtLength('M0,0 L10,10')

    const request = [
      0, 0.01, 1, 1.01, 2, 2.01, 3, 3.01, 4, 4.01, 5, 5.01, 6, 6.01, 7, 7.01, 8,
      8.01, 9, 9.01, 10, 10.01, 11, 11.01, 12, 12.01, 13, 13.01, 14, 14.01,
      14.142135623730951,
    ]

    const indexedResult = indexed.atBatch(request).map((p) => p.pos)
    const originalResult = request.map((len) => indexed.at(len))

    expect(indexedResult).toEqual(originalResult)
  })

  it('processInput', async () => {
    const result = await processInput({
      id: 'test',
      type: 'getPoints',
      brushSize: 1,
      destSize: { width: 10, height: 10 },
      inOutInfluence: 0,
      scatterRange: 0,
      scatterScale: 0,
      inOutLength: 0,
      path: {
        randomSeed: 0,
        closed: false,
        points: [
          {
            x: 0,
            y: 0,
            end: null,
            begin: null,
          },
          {
            x: 10,
            y: 10,
            begin: null,
            end: null,
          },
        ],
      },
    })

    result.type === 'getPoints' && console.log(result._internals)

    // expect(result).toMatchSnapshot()
  })
})
