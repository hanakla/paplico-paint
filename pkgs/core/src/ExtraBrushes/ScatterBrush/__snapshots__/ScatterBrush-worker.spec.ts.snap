// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScatterBrush-worker > processInput 1`] = `
{
  "_debug": {
    "path": {
      "fillRule": "nonzero",
      "points": [
        {
          "begin": null,
          "end": null,
          "isMoveTo": true,
          "x": 0,
          "y": 0,
        },
        {
          "begin": null,
          "end": null,
          "x": 10,
          "y": 10,
        },
      ],
      "randomSeed": 0,
    },
    "positions": [
      [
        0,
        0,
      ],
      [
        0.7071067811865475,
        0.7071067811865475,
      ],
      [
        1.414213562373095,
        1.414213562373095,
      ],
      [
        2.121320343559642,
        2.121320343559642,
      ],
      [
        2.82842712474619,
        2.82842712474619,
      ],
      [
        3.5355339059327378,
        3.5355339059327378,
      ],
      [
        4.242640687119286,
        4.242640687119286,
      ],
      [
        4.949747468305834,
        4.949747468305834,
      ],
      [
        5.65685424949238,
        5.65685424949238,
      ],
      [
        6.363961030678926,
        6.363961030678926,
      ],
      [
        7.071067811865471,
        7.071067811865471,
      ],
      [
        7.778174593052015,
        7.778174593052015,
      ],
      [
        8.485281374238564,
        8.485281374238564,
      ],
      [
        9.192388155425112,
        9.192388155425112,
      ],
      [
        9.89949493661166,
        9.89949493661166,
      ],
      [
        10,
        10,
      ],
    ],
    "requestAts": [
      0,
      0.01,
      1,
      1.01,
      2,
      2.01,
      3,
      3.01,
      4,
      4.01,
      5,
      5.01,
      6,
      6.01,
      7,
      7.01,
      8,
      8.01,
      9,
      9.01,
      10,
      10.01,
      11,
      11.01,
      12,
      12.01,
      13,
      13.01,
      14,
      14.01,
      14.142135623730958,
    ],
  },
  "bbox": {
    "bottom": 10,
    "left": 0,
    "right": 10,
    "top": 0,
  },
  "id": "test",
  "lengths": [
    0,
    1.1313708498984762,
    2.121320343559643,
    3.11126983722081,
    4.101219330881976,
    5.091168824543141,
    6.081118318204308,
    7.071067811865475,
    8.06101730552664,
    9.050966799187812,
    10.04091629284898,
    11.030865786510152,
    12.020815280171316,
    13.010764773832484,
    14.000714267493649,
    14.142135623730958,
  ],
  "matrices": [
    [
      0.8216791726206834,
      0.5699502936935722,
      0,
      0,
      -0.5699502936935722,
      0.8216791726206834,
      0,
      0,
      0,
      0,
      1,
      0,
      -5,
      5,
      0,
      1,
    ],
    [
      -0.3556864836728372,
      0.9346053313203667,
      0,
      0,
      -0.9346053313203667,
      -0.3556864836728372,
      -0,
      -0,
      0,
      0,
      1,
      0,
      -4.292893218813452,
      4.292893218813452,
      0,
      1,
    ],
    [
      0.0270291671131038,
      -0.999634645320565,
      0,
      0,
      0.999634645320565,
      0.0270291671131038,
      0,
      0,
      0,
      0,
      1,
      0,
      -3.585786437626905,
      3.585786437626905,
      0,
      1,
    ],
    [
      0.43302900828369817,
      0.9013799853473766,
      0,
      0,
      -0.9013799853473766,
      0.43302900828369817,
      0,
      0,
      0,
      0,
      1,
      0,
      -2.878679656440358,
      2.878679656440358,
      0,
      1,
    ],
    [
      -0.6280973688545088,
      0.7781347539070871,
      0,
      0,
      -0.7781347539070871,
      -0.6280973688545088,
      -0,
      -0,
      0,
      0,
      1,
      0,
      -2.17157287525381,
      2.17157287525381,
      0,
      1,
    ],
    [
      -0.16083405590121091,
      0.9869814620662164,
      0,
      0,
      -0.9869814620662164,
      -0.16083405590121091,
      -0,
      -0,
      0,
      0,
      1,
      0,
      -1.4644660940672622,
      1.4644660940672622,
      0,
      1,
    ],
    [
      0.9985086728795873,
      0.05459331629645908,
      0,
      0,
      -0.05459331629645908,
      0.9985086728795873,
      0,
      0,
      0,
      0,
      1,
      0,
      -0.7573593128807143,
      0.7573593128807143,
      0,
      1,
    ],
    [
      0.709333001018619,
      -0.7048735302633514,
      0,
      0,
      0.7048735302633514,
      0.709333001018619,
      0,
      0,
      0,
      0,
      1,
      0,
      -0.050252531694166436,
      0.050252531694166436,
      0,
      1,
    ],
    [
      -0.8065824034996864,
      -0.5911216679877919,
      -0,
      -0,
      0.5911216679877919,
      -0.8065824034996864,
      0,
      0,
      0,
      0,
      1,
      0,
      0.6568542494923797,
      -0.6568542494923797,
      0,
      1,
    ],
    [
      -0.8735573320503071,
      -0.48672126275841854,
      -0,
      -0,
      0.48672126275841854,
      -0.8735573320503071,
      0,
      0,
      0,
      0,
      1,
      0,
      1.3639610306789258,
      -1.3639610306789258,
      0,
      1,
    ],
    [
      0.952787774644259,
      0.3036370472923895,
      0,
      0,
      -0.3036370472923895,
      0.952787774644259,
      0,
      0,
      0,
      0,
      1,
      0,
      2.071067811865471,
      -2.071067811865471,
      0,
      1,
    ],
    [
      -0.7323956918296335,
      0.6808792481706228,
      0,
      0,
      -0.6808792481706228,
      -0.7323956918296335,
      -0,
      -0,
      0,
      0,
      1,
      0,
      2.7781745930520154,
      -2.7781745930520154,
      0,
      1,
    ],
    [
      0.2770347376086805,
      -0.9608599035021128,
      0,
      0,
      0.9608599035021128,
      0.2770347376086805,
      0,
      0,
      0,
      0,
      1,
      0,
      3.485281374238564,
      -3.485281374238564,
      0,
      1,
    ],
    [
      0.9544294141958162,
      0.2984367492750033,
      0,
      0,
      -0.2984367492750033,
      0.9544294141958162,
      0,
      0,
      0,
      0,
      1,
      0,
      4.192388155425112,
      -4.192388155425112,
      0,
      1,
    ],
    [
      -0.9959549674545818,
      0.08985378568843358,
      0,
      0,
      -0.08985378568843358,
      -0.9959549674545818,
      -0,
      -0,
      0,
      0,
      1,
      0,
      4.89949493661166,
      -4.89949493661166,
      0,
      1,
    ],
    [
      -0.7496596482539286,
      0.6618235503363386,
      0,
      0,
      -0.6618235503363386,
      -0.7496596482539286,
      -0,
      -0,
      0,
      0,
      1,
      0,
      5,
      -5,
      0,
      1,
    ],
  ],
  "totalLength": 14.142135623730958,
  "type": "getPoints",
}
`;
